### DISCUSSION TITLE
[RFC] ConfigProvider support `animation={false}`

### DISCUSSION THREAD

**Discussion by afc163** [ant-design/ant-design#40040] [RFCs - archive]
## 概况

ConfigProvider 新增 `animation` 属性，用于开启/关闭所有组件的动画效果。

## Motivation

- 在某些复杂业务场景下动画效果会影响性能。如 https://github.com/ant-design/ant-design/issues/25698
- 部分场景需要极致快捷的交互操作，部分弹层的动画效果会显得迟滞。https://www.jianshu.com/p/c1c2cdaa2af0

参考 https://ng.ant.design/docs/animations/zh 可以全局关闭组件动画效果。

## API

**ConfigProvider**

| 参数 | 说明 | 类型 | 默认值 | 
| --- | --- | --- | --- | 
| animation | 组件全局配置 | Boolean | - | 

## Basic Example

```tsx
import { ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider animation={false}>
    ...
  </ConfigProvider>
)
```

## Detailed

- [ ] 把 Context 传入给 design token 生成逻辑，修改对应的 transition 和 animation 变量。
- [ ] 补充相关文档，作为最佳实践之一。

## Drawbacks

需要修改有动画的组件实现，有一定侵入性。

## Other solutions

下面这个也是个可行方案，而且无需开发现在就能用，不过要检查一下能不能覆盖所有动画效果。

```tsx
import { ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider theme={{
    token: {
      motionDurationFast: 0,
      motionDurationMid: 0,
      motionDurationSlow: 0,
    },
  }}>
    ...
  </ConfigProvider>
)
```

**Comment by dancerphil** [comment]
## 补充一个 Motivation

3. 覆盖 animation，如果可以关闭，在覆盖 animation 时会比较方便

## 对应的 API 的想法

按维度开关：比如 Wave 和 Modal 的展开收起是否可以分别开关。

**Comment by arvinxx** [comment]
token的覆盖应该用 `motionUnit` 这个。

```tsx
import { ConfigProvider } from 'antd';

export default () => (
  <ConfigProvider theme={{
    token: {
      motionUnit: 0,
    },
  }}>
    ...
  </ConfigProvider>
)
```

fast、mid、slow都是基于unit派生的：

```ts
{
 // motion
    motionDurationFast: `${(motionBase + motionUnit).toFixed(1)}s`,
    motionDurationMid: `${(motionBase + motionUnit * 2).toFixed(1)}s`,
    motionDurationSlow: `${(motionBase + motionUnit * 3).toFixed(1)}s`,
}
```

**Reply by afc163** [reply]
motionUnit 感觉语义上不太能看懂 unit 的具体单位是啥？

**Reply by arvinxx** [reply]
可以再改改，比如叫motionSpeedUnit 估计会更加清楚点。 现在的单位应该是秒，方便计算。我感觉是不是可以把ms的配置也支持上？比如读取到 "50ms" 背后自动转成0.05。 如果不是ms或者s结尾的其他字符串就抛错误

**Comment by arvinxx** [comment]
@afc163 @zombieJ  @MadCcc  最近在开发中遇到了一个问题，使用 motionUnit =0.001 无法解决。

测试了 CodeSandbox，是正常的：https://codesandbox.io/s/modal-qu-dong-hua-gong-neng-np6h6q?file=/demo.tsx:0-195

但如果使用 nextjs，则会有问题（点两次就能复现）：https://codesandbox.io/p/sandbox/proud-sun-dbhceq
